pkg/operator/controller/route-metrics/controller.go

```diff
@@ -172,55 +163,26 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
-
 	// List all the Namespaces filtered by our ingress's Namespace selector.
-	namespaceMatchingLabelsSelector := client.MatchingLabelsSelector{Selector: labels.Everything()}
-	if ingressController.Spec.NamespaceSelector != nil {
-		namespaceSelector, err := metav1.LabelSelectorAsSelector(ingressController.Spec.NamespaceSelector)
-		if err != nil {
-			log.Error(err, "ingresscontroller has an invalid namespace selector", "ingresscontroller",
-				ingressController.Name, "namespaceSelector", ingressController.Spec.NamespaceSelector)
-			return reconcile.Result{}, nil
-		}
-		namespaceMatchingLabelsSelector = client.MatchingLabelsSelector{Selector: namespaceSelector}
-	}
-
-	namespaceList := corev1.NamespaceList{}
-	if err := r.cache.List(ctx, &namespaceList, namespaceMatchingLabelsSelector); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %w", request, err)
-	}
-	// Create a set of Namespaces to easily look up Namespaces that matches the Routes assigned to the Ingress Controller.
-	namespacesSet := sets.NewString()
-	for i := range namespaceList.Items {
-		namespacesSet.Insert(namespaceList.Items[i].Name)
+	namespacesInShard, err := routestatus.GetNamespacesSelectedByIngressController(r.cache, ingressController)
+	if err != nil {
+		return reconcile.Result{}, err
 	}
 
-	// List routes filtered by our ingress's route selector.
-	routeMatchingLabelsSelector := client.MatchingLabelsSelector{Selector: labels.Everything()}
-	if ingressController.Spec.RouteSelector != nil {
-		routeSelector, err := metav1.LabelSelectorAsSelector(ingressController.Spec.RouteSelector)
-		if err != nil {
-			log.Error(err, "ingresscontroller has an invalid route selector", "ingresscontroller",
-				ingressController.Name, "routeSelector", ingressController.Spec.RouteSelector)
-			return reconcile.Result{}, nil
-		}
-		routeMatchingLabelsSelector = client.MatchingLabelsSelector{Selector: routeSelector}
-	}
-	routeList := routev1.RouteList{}
-	if err := r.cache.List(ctx, &routeList, routeMatchingLabelsSelector); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to list Routes for the Shard %q: %w", request, err)
+	// List all the Namespaces filtered by our ingress's Namespace selector.
+	routesInShard, err := routestatus.GetRoutesSelectedByIngressController(r.cache, ingressController)
+	if err != nil {
+		return reconcile.Result{}, err
 	}
 
 	// Variable to store the number of routes admitted by the Shard (Ingress Controller).
 	routesAdmitted := 0
 
 	// Iterate through the list Routes.
-	for _, route := range routeList.Items {
+	for _, route := range routesInShard.Items {
 		// Check if the Route's Namespace matches one of the Namespaces in the set namespacesSet and
 		// the Route is admitted by the Ingress Controller.
-		if namespacesSet.Has(route.Namespace) && routeStatusAdmitted(route, ingressController.Name) {
+		if namespacesInShard.Has(route.Namespace) {
```

Comment:
I think you missed out calling the `IsRouteStatusAdmitted` function here.
```suggestion
		if namespacesInShard.Has(route.Namespace) && routestatus.IsRouteStatusAdmitted(route, ingressController.Name) {
```

Reply:
Thanks for the review (we were going to actually ping you to review this).

Yea I debated this, dropped it intentionally, and wanted to have a discussion. I think our options are:
1. Just check route/namespace label matching (my current revision)
2. Check route/namespace label matching AND route status (current master)
3. Check just route status

I forgot about option 3 until now - since we are actually fixing route status in the PR, technically we could just call `IsRouteStatusAdmitted` solely, without doing ANY lookups, which I think is much more efficient, but doesn't protect us against bugs in the future (not sure if that's a valid argument though, bugs should be fixed, and other things rely solely on route status like external dns).

I'm leaning toward option 3, let me know what you think or if I forgot anything.

Reply:
> 1. Just check route/namespace label matching (my current revision)

Is the route guaranteed to be admitted (immediately) by the Ingress if the route/namespace label match? If so, then checking the route status would not be required. If not (due to delay, misconfiguration, failure, etc.), then route metrics would be incorrect. I am not sure about how route admission works, so leaving this open.

> 2. Check route/namespace label matching AND route status (current master)

Out of the 3, this option as of now looks optimal to me as we are filtering out the route first and then also checking the status to ensure the route metrics are in sync.

> 3. Check just route status

Are all the routes going to be fetched in this scenario and then for each route the `IsRouteStatusAdmitted` function will be called? If so, then aren't we adding a lot of additional work to check the status of the routes which could be easily filtered out by the labelselectors? This would add a lot of load to ingress operator if a cluster has a large number of routes.

Let me know if I have understood the 3 scenarios correctly.

Reply:
I kind of approximated the time complexity of both the approaches: 1) With namespace filtering, 2) Without namespace filtering. Not sure if the assumptions are correct, but here it goes.

1\) Let's assume that there are `n` namespaces filtered out by the namespace labelselector. Assuming that fetching the namespaces from the k8s api is dependent on the number of namespaces being fetched, then the time complexity will be `O(n)`. After the namespaces are fetched, each namespace is iterated over to construct the Set, the time complexity of which is again `O(n)`. Now, let there be `p+q` routes fetched using the `routeSelector`. `p` routes don't match the `namespaceSelector` and `q` routes match it. Let's assume that for each route there are `r` ingresses in the status. Then, following is the time complexity of the below code-snippet:
```go
// Iterate through the list Routes.
for _, route := range routesInShard.Items {
// Check if the Route's Namespace matches one of the Namespaces in the set namespacesSet and
// the Route is admitted by the Ingress Controller.
if namespacesInShard.Has(route.Namespace) && routestatus.IsRouteStatusAdmitted(route, ingressController.Name) {
	// If the Route is admitted then, the routesAdmitted should be incremented by 1 for the Shard.
	routesAdmitted++
}
```
The namespace filtering will be done for all routes, so the time complexity will be `O(p+q)` as the implementation of the Set is done using hashmaps and the lookup in a hashmap is `O(1)`. The check for route's admitted status will run for only `q` routes and the time complexity will be `O(qr)`.
So, the overall time complexity for this approach will be: `O(n) + O(n) + O(p+q) + O(qr)`. This will boil down to `O(qr)` as the lower order time complexity will not have as much impact as the higher order time complexity. Though I am not completely sure whether the time complexity of fetching the namespaces should be `O(n)`.

2\) As there are no namespace filtering, the time complexity calculation will only involve calculating it for the following code snippet:
  ```go
  // Iterate through the list Routes.
  for _, route := range routesInShard.Items {
  // Check if the Route's Namespace matches one of the Namespaces in the set namespacesSet and
  // the Route is admitted by the Ingress Controller.
  if routestatus.IsRouteStatusAdmitted(route, ingressController.Name) {
	  // If the Route is admitted then, the routesAdmitted should be incremented by 1 for the Shard.
	  routesAdmitted++
  }
  ```
Going with the same assumption about the number routes and ingresses per route, the time complexity for checking the route's admitted status will be `O(pr+qr)` as it is executed for all the `p+q` routes.

So, looking at the time complexities of both the approaches the factor of `O(pr)` is the additional impact we get for not filtering out the routes using the `namespaceSelector`. For clusters where the number of routes are small the benefit might not be significant. In fact, the extra work done for namespace filtering (fetching and constructing the Set) might sometimes (but not always) result in a little bit higher time. Whereas, for clusters with large routes the namespace filtering will beneficial as we are avoiding running through the nested loop for the filtered out routes. Again, all the calculations depend on the assumption of the time complexity for fetching the namespaces from the k8s api.

Thoughts?

Reply:
Great analysis, I'm a little rusty on O notation, but I appreciate you taking this analytical approach and keeping me honest.

As written, yes I think you have a valid argument. Did you mean an exponential here: `O(qʳ)` and `O(pʳ)`?

> Though I am not completely sure whether the time complexity of fetching the namespaces should be `O(n)`.

I think it's a little more than just O(n). Looking at the list code: https://github.com/kubernetes-sigs/controller-runtime/blob/ff622bd6042d2ebecb5cfc1dcf8c5b63cf47ec5c/pkg/cache/internal/cache_reader.go#L107

1. List all namespaces and filter by label selector. we'll call N all namespaces `O(N)`
2. Return filtered namespaces.

Then, we build a set from these I think that's `O(n)`. So wouldn't this be `O(N+n)`? But I don't think this matters because this function boils down to the highly complexity of `O(qʳ)`. 

Regardless, I think you have a valid point for efficiency here regarding filtering. I think we still want to check the route status regardless to make sure there is feedback from the router pod that it actually admitted the route (that's why I need to put back `IsRouteStatusAdmitted`).

De-duplicating logic was my priority (especially since https://issues.redhat.com/browse/OCPBUGS-2848), then efficiency next, but I think you have a good point about efficient. I just wonder if there is a nuance we are missing, but for the time being, I'll put it back the way it was.

Sorry for this tangent, but thanks for responding.

Reply:
> As written, yes I think you have a valid argument. Did you mean an exponential here: `O(q)` and `O(pʳ)`?

`O(pr)` and `O(qr)` are simply `O(p*r)` and `O(q*r)` which can be interpreted as `O(n^2)`.

> 1. List all namespaces and filter by label selector. we'll call N all namespaces `O(N)`
> 2. Return filtered namespaces.
> 
> Then, we build a set from these I think that's `O(n)`. So wouldn't this be `O(N+n)`?

We can take the upper limit and consider it as `O(2*N)` which is same as `O(N)` as the constant 2 would not have much of an effect. But I was mostly referring to the overhead of the RPC which I am not sure about.

> De-duplicating logic was my priority (especially since https://issues.redhat.com/browse/OCPBUGS-2848), then efficiency next, but I think you have a good point about efficient. I just wonder if there is a nuance we are missing, but for the time being, I'll put it back the way it was.

Maybe we can think of a different way of implementing the route metrics controller. We can do something like this (https://github.com/openshift/cluster-ingress-operator/pull/779/commits/04df4b7994bb6db3eecd048c065087e47637b32c). Though this implementation makes a lot of assumptions about how `predicate.Funcs` work and valid questions were raised (https://github.com/openshift/cluster-ingress-operator/pull/779#discussion_r939069420). But the same logic can be implemented without using them as well. Here we only update the metric for the particular route for which a CRUD event happened. We just depend on the route status of the particular.

Reply:
> O(pr) and O(qr) are simply O(p*r) and O(q*r) which can be interpreted as O(n^2).

Okay yea you are right, I had to brush up on O Notation.

> Maybe we can think of a different way of implementing the route metrics controller. We can do something like this (https://github.com/openshift/cluster-ingress-operator/commit/04df4b7994bb6db3eecd048c065087e47637b32c). Though this implementation makes a lot of assumptions about how predicate.Funcs work and valid questions were raised (https://github.com/openshift/cluster-ingress-operator/pull/779#discussion_r939069420). But the same logic can be implemented without using them as well. Here we only update the metric for the particular route for which a CRUD event happened. We just depend on the route status of the particular route.

Yea that's a valid point and I'll keep that pattern in mind, but I think I'm going to abstain from further improvements to the route-metrics controller in this PR and we can rewrite if it efficiency issues come up later with the route-metrics controller in a different PR. I'll keep my PRs modifications to make logic common between route-metric and route-status for the time being.

Reply:
> I think I'm going to abstain from further improvements to the route-metrics controller in this PR and we can rewrite if it efficiency issues come up later with the route-metrics controller in a different PR

Makes sense. That can be taken up separately whenever required.